VisualRiders.PointOfSale.Project/Services/DiscountsService.cs

```diff
@@ -0,0 +1,195 @@
+﻿using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore.ValueGeneration.Internal;
+using System.Diagnostics.Tracing;
+using VisualRiders.PointOfSale.Project.DTOs;
+using VisualRiders.PointOfSale.Project.Exceptions;
+using VisualRiders.PointOfSale.Project.Models;
+using VisualRiders.PointOfSale.Project.Repositories;
+
+namespace VisualRiders.PointOfSale.Project.Services;
+
+public class DiscountsService
+{
+    private readonly DiscountsRepository _repository;
+    private readonly DiscountItemsRepository _discountItemsRepository;
+    private readonly ProductsRepository _productsRepository;
+    private readonly ServicesRepository _servicesRepository;
+    private readonly IMapper _mapper;
+
+    public DiscountsService(
+        DiscountsRepository repository,
+        DiscountItemsRepository discountItemsRepository,
+        ProductsRepository productsRepository,
+        ServicesRepository servicesRepository,
+        IMapper mapper)
+    {
+        _repository = repository;
+        _discountItemsRepository = discountItemsRepository;
+        _productsRepository = productsRepository;
+        _servicesRepository = servicesRepository;
+        _mapper = mapper;
+    }
+
+    public Discount Create(CreateUpdateDiscountDto dto)
+    {
+        var discount = _mapper.Map<Discount>(dto);
+
+        discount.BusinessEntityId = 1;
+        discount.DiscountItems = new List<DiscountItem>();
+        
+        _repository.Add(discount);
+        _repository.SaveChanges();
+
+        return discount;
+    }
+
+    public List<ReadDiscountDto> GetAll()
+    {
+        return _repository.GetAll().Select(_mapper.Map<ReadDiscountDto>).ToList();
+    }
+
+    public ReadDiscountDto? GetById(int id)
+    {
+        return _mapper.Map<ReadDiscountDto>(_repository.GetById(id));
+    }
+
+    public ReadDiscountDto? UpdateById(int id, CreateUpdateDiscountDto dto)
+    {
+        var discount = _repository.GetById(id);
+
+        if (discount == null) return null;
+
+        discount = _mapper.Map<Discount>(dto); 
+
+        _repository.SaveChanges();
+
+        return _mapper.Map<ReadDiscountDto>(discount);
+    }
+
+    public ReadDiscountItemDto? AddItemById(int id, CreateUpdateDiscountItemDto dto)
```

Comment:
The case where both `DiscountId` and `ServiceId` are null is not handled. Also, it is illegal for both of them to be set as one discountItem is allowed to reference one product or one service. Lastly, you did not handle the case where the item already has the discount applied, so it would be possible to have duplicate items for the same discount with different discount sizes and it would not be possible to determine, which discount to apply.